apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClientImpl.DeleteBucket.cs

```diff
@@ -23,37 +24,79 @@ public sealed partial class StorageClientImpl : StorageClient
 
-        private BucketsResource.DeleteRequest CreateDeleteBucketRequest(Bucket bucket, DeleteBucketOptions options)
+        // Implementation methods assuming a previously-validate bucket name
+
+        private void DeleteBucketImpl(string bucket, DeleteBucketOptions options)
         {
-            ValidateBucket(bucket, nameof(bucket));
-            var request = Service.Buckets.Delete(bucket.Name);
-            options?.ModifyRequest(request);
-            return request;
+            // Create the request at the start, so we validate the options before we
+            // start deleting objects.
+            var deleteBucketRequest = CreateDeleteBucketRequest(bucket, options);
+            string userProject = options?.UserProject;
+            if (options?.DeleteObjects ?? false)
+            {
+                // Don't start deleting objects if the bucket has the wrong metageneration to start with.
+                var preconditionOptions = options?.CreateGetBucketOptionsForPreconditions();
+                if (preconditionOptions != null)
+                {
+                    GetBucket(bucket, preconditionOptions);
+                }
+                foreach (var obj in ListObjects(bucket, null, new ListObjectsOptions { Versions = true, UserProject = userProject }))
+                {
+                    var deleteObjectOptions = new DeleteObjectOptions { UserProject = userProject, Generation = obj.Generation };
+                    DeleteObject(obj, deleteObjectOptions);
+                }
+            }
+            deleteBucketRequest.Execute();
+        }
+
+        private async Task DeleteBucketAsyncImpl(string bucket, DeleteBucketOptions options, CancellationToken cancellationToken)
+        {
+            var deleteBucketRequest = CreateDeleteBucketRequest(bucket, options);
+            string userProject = options?.UserProject;
+            if (options?.DeleteObjects ?? false)
+            {
+                // Don't start deleting objects if the bucket has the wrong metageneration to start with.
+                var preconditionOptions = options?.CreateGetBucketOptionsForPreconditions();
+                if (preconditionOptions != null)
+                {
+                    await GetBucketAsync(bucket, preconditionOptions, cancellationToken).ConfigureAwait(false);
+                }
+                var objects = ListObjectsAsync(bucket, null, new ListObjectsOptions { Versions = true, UserProject = userProject });
+                using (var iterator = objects.GetEnumerator())
+                {
+                    while (await iterator.MoveNext(cancellationToken).ConfigureAwait(false))
+                    {
+                        var obj = iterator.Current;
+                        var deleteObjectOptions = new DeleteObjectOptions { UserProject = userProject, Generation = obj.Generation };
+                        await DeleteObjectAsync(obj, deleteObjectOptions, cancellationToken).ConfigureAwait(false);
```

Comment:
Shouldn't we collect all of these into a list, and then to `await Task.WhenAll(taskList)`? Is there a concern about the number of open connections, or does the client self throttle?

Reply:
Task[] deleteTasks = await objects.Select(
        obj =>
        {
            var deleteObjectOptions =
                new DeleteObjectOptions {UserProject = userProject, Generation = obj.Generation};
            return DeleteObjectAsync(obj, deleteObjectOptions, cancellationToken);
        }).ToArray(cancellationToken).ConfigureAwait(false);
    await Task.WhenAll(deleteTasks).ConfigureAwait(false);

Reply:
I definitely don't want to do it in an unconstrained way. There can be millions of objects in a bucket. If wanted to parallelize, I'd potentially use `AsRawResponses` and delete a page of data at a time (which is 1000 objects by default), and even then possibly set concurrency to ~8 concurrent requests.